[move] Update why move (#452)
aptos-labs/developer-docs: apps/nextra/pages/en/build/smart-contracts/why-move.mdx

Move was built upon the idea that other programming languages made it harder to
provide guarantees at runtime, which is required for a trusted execution
environment.

Move is built upon the following principals:

1. **Secure by default**: Financial systems are built to ensure users don't lose
funds.  Move was designed to prevent entire classes of attacks and bugs such as
reentrancy attacks, double spends, and arithmetic overflow.  Type safety and
compile time checks are at the forefront of the security.

2. **Runtime Verification**: The bytecode can be verified at runtime to verify
that nothing has gone wrong, providing extra safety and preventing malicious actors.

3. **Formal Verification**: The language provides an extension to the Move language
specifically to provide formal verification of contracts.  This allows for proving
invariants and as a tool to help code auditing.

4. **Simplicity**: The commands and bytecode are purposely simple.  This allows
for easy decompilation, runtime verification, and code inspection.  More complex
languages often have whole parts that need to be ignored e.g. Rust.

# Why Move on Aptos?

> Move on Aptos supports the full language built by the team at Facebook, with some

Reviewer: s/some additional/additional

an easy-to-parse format.

3. **Upgradability**: Upgradability ensures that application interfaces cannot
be broken and doesn't require explicit adoption from downstream applications.
Contracts can simply be upgraded in-place to fix bugs.

4. **Cross-Interaction**: Move allows for interaction between contracts by using
type-safe structs.

5. **Code Storage**: Aptos stores source code on-chain improving the ability to
audit and ensure contract to bytecode correctness.

6. **Sponsored Transactions**: Native sponsored transaction support allows for
having transactions be paid by other users with no special services or
contract-specific code required.

7.  **Robust Token Standards**: The Digital Asset and Fungible Asset standards
provide flexibility and a unified standard for diverse types of tokens and
digital assets on-chain.  These were influenced by existing standards such as
ERC-20, ERC-721, ERC-1155 and Token-2022.

8. **On-chain Randomness**: Native on-chain unbiasable randomness provides a safe
and consistent way of getting random numbers, with extra safety checks at compile
time.

## Comparison to Other Blockchains

> 

Reviewer: I thinks its fine right now, but do not promise anything about not having dynamic dispatch (no re-entrency is fine though)

# Why Move?

The Move programming language was originally created by a team of engineers at
Facebook for the Diem Payment Network. Move is designed to be a platform-agnostic
language to enable common libraries, tooling, and developer communities across
diverse blockchains with vastly different data and execution models.
At Aptos, we believe in building a strong developer community in Move and invite
them to build upon the Move on Aptos stack and contribute to the open source software.

> Move was built upon the idea that other programming languages made it harder to
> provide guarantees at runtime, which is required for a trusted execution
> environment.

Reviewer 2: I'm a bit confused by what you mean here. Some clarity or examples on the guarantees might help. Maybe something like:

```suggestion
Move was designed to provide safety and security guarantees at runtime, a novel idea not implemented by
other smart contract languages.
```

Probably want to add something else after

Author: I'm removing this sentence

Move is built upon the following principals:

1. **Secure by default**: Financial systems are built to ensure users don't lose
funds.  Move was designed to prevent entire classes of attacks and bugs such as
reentrancy attacks, double spends, and arithmetic overflow.  Type safety and
compile time checks are at the forefront of the security.

2. **Runtime Verification**: The bytecode can be verified at runtime to verify
that nothing has gone wrong, providing extra safety and preventing malicious actors.

> 3. **Formal Verification**: The language provides an extension to the Move language
> specifically to provide formal verification of contracts.  This allows for proving

Reviewer 2: ```suggestion
3. **Formal Verification**: Move on Aptos includes an extension to the language 
-- the Move Prover -- to provide formal verification of contracts.  This allows for proving
```

> invariants and as a tool to help code auditing.

Reviewer 2: ```suggestion
invariants and helps with code auditing.
```

4. **Simplicity**: The commands and bytecode are purposely simple.  This allows
for easy decompilation, runtime verification, and code inspection. Using regular
programming languages for blockchains often requires to ignore large part of the
language to make them suitable for smart contracts (e.g. Rust).

# Why Move on Aptos?

Move on Aptos supports the full language built by the team at Facebook, with
additional extensions built to improve the security and the developer experience.

## Security

1. **Formal Verification**: Aptos framework is fully specified and formally

> verified. This includes the core contracts involving governance, NFTs, and Tokens.

Reviewer 2: ```suggestion
verified via the Move Prover. This includes the core contracts involving governance, NFTs, and Tokens.
```

2. **Gas Coverage**: Move VM has 100% gas coverage. Gas is charged based upon
actual usage in the system (CPU, memory, storage, I/O). In other words, no
gas exploits.

3. **Security Redundancy**: Security redundancy provided by runtime safety checks.

4. **Permission Controls**: Permission controls can flexibly be built at various

> levels.  For example, at the token level permission controls are by default to

Reviewer 2: I think you're missing something here. `...are __blank__ by default...`

enable RWA tokenization.

## Developer Experience

1. **Move Development Tools**:

>   - **Unit testing**: Move has built in unit testing for all contracts.  Aptos

Reviewer 2: ```suggestion
  - **Unit testing**: Move has built-in unit testing for all contracts.  Aptos
```

> additionally, provides test functionality in the framework to test different

Reviewer 2: ```suggestion
additionally provides test functionality in the framework to test different
```
